nixos/modules/services/web-servers/ttyd.nix

```diff
@@ -0,0 +1,202 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.ttyd;
+
+  # Command line arguments for the ttyd daemon
+  args = [ "--port ${toString cfg.port}" ]
+         ++ optional (cfg.socket != null) "--interface ${cfg.socket}"
+         ++ optional (cfg.interface != null) "--interface ${cfg.interface}"
+         ++ optional (cfg.username != null) "--credential ${escapeShellArg (cfg.username + ":" + cfg.password)}"
+         ++ optional (cfg.uid != null) "--uid ${cfg.uid}"
+         ++ optional (cfg.gid != null) "--gid ${cfg.gid}"
+         ++ optional (cfg.signal != null) "--signal ${toString cfg.signal}"
+         ++ (mapAttrsToList (_k: _v: "--client-option ${_k}=${escapeShellArg _v}") cfg.clientOptions)
+         ++ optional (cfg.terminalType != null) "--terminal-type ${cfg.terminalType}"
+         ++ optional cfg.checkOrigin "--check-origin"
+         ++ optional (cfg.maxClients != null) "--max-clients ${cfg.maxClients}"
+         ++ optional (cfg.indexFile != null) "--index ${cfg.indexFile}"
+         ++ optional cfg.enableIPv6 "--ipv6"
+         ++ optional cfg.enableSSL "--ssl-cert ${cfg.certFile} --ssl-key ${cfg.keyFile} --ssl-ca ${cfg.caFile}"
+         ++ optional (cfg.logLevel != null) "--debug ${toString cfg.logLevel}"
+
+  ;
+
+  # Command to start ttyd
+  cmd = "${pkgs.ttyd}/bin/ttyd ${concatStringsSep " " args} ${pkgs.shadow}/bin/login";
```

Comment:
Use `lib.escapeShellArgs` instead of `concatStringsSep " "`

Reply:
Oh wow, this is so much better! Thanks. Fixed.